store/sqlstore/plugin_store.go

```diff
@@ -108,6 +109,33 @@ func (ps SqlPluginStore) CompareAndSet(kv *model.PluginKeyValue, oldValue []byte
 			// Failed to update
 			return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.save.app_error", nil, err.Error(), http.StatusInternalServerError)
 		} else if rowsAffected == 0 {
+			if ps.DriverName() == model.DATABASE_DRIVER_MYSQL && bytes.Equal(oldValue, kv.Value) {
+				// ROW_COUNT on MySQL is zero even if the row existed but no changes to the row were required.
+				// Check if the row exists with the required value to distinguish this case. Strictly speaking,
+				// this isn't a good use of CompareAndSet anyway, since there's no corresponding guarantee of
+				// atomicity. Nevertheless, let's return results consistent with Postgres and with what might
+				// be expected in this case.
+				count, err := ps.GetReplica().SelectInt(
+					"SELECT COUNT(*) FROM PluginKeyValueStore WHERE PluginId = :PluginId AND PKey = :Key AND PValue = :Value",
+					map[string]interface{}{
+						"PluginId": kv.PluginId,
+						"Key":      kv.Key,
+						"Value":    kv.Value,
+					},
+				)
+				if err != nil {
+					return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.compare_and_set.mysql_select.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, err=%v", kv.PluginId, kv.Key, err.Error()), http.StatusInternalServerError)
+				}
+
+				if count == 0 {
+					return false, nil
+				} else if count == 1 {
+					return true, nil
+				} else {
+					return false, model.NewAppError("SqlPluginStore.CompareAndSet", "store.sql_plugin_store.compare_and_set.too_many_rows.app_error", nil, fmt.Sprintf("plugin_id=%v, key=%v, count=%d", kv.PluginId, kv.Key, count), http.StatusInternalServerError)
```

Comment:
Just to double check: This should _never_ happen, right?

Reply:
Correct: the primary keys, if in place, guarantee the database would never allow this situation to exist.